libs/Microsoft.MixedReality.WebRTC.Unity/Assets/Microsoft.MixedReality.WebRTC.Unity/Scripts/Media/RemoteAudioSource.cs

```diff
@@ -173,9 +332,73 @@ private void TrackRemoved(WebRTC.PeerConnection.TrackKind trackKind)
             }
         }
 
-        //private void RemoteAudioFrameReady(AudioFrame frame)
-        //{
-        //    FrameQueue.Enqueue(frame);
-        //}
+        private void EnqueueCreateAudioClipOnMainThread(int sampleRate, int channelCount)
+        {
+            Debug.Assert(_audioClipEnqueuedCreate == false);
+            _audioClipEnqueuedCreate = true;
+            _mainThreadWorkQueue.Enqueue(() =>
+            {
+                int lengthSamples = sampleRate / 10; // Buffer up to 100ms of data.
+                _audioClip = AudioClip.Create("RemoteClip", sampleRate, channelCount, sampleRate, true, _audioData.OnClipRead);
+                _audioClipEnqueuedCreate = false;
+                var aud = GetComponent<UnityEngine.AudioSource>();
+                if (aud)
+                {
+                    aud.clip = _audioClip;
+                    aud.volume = 1.0f;
+                    aud.loop = true;
+                    aud.bypassEffects = true;
+                    aud.Play();
+                }
+            });
+        }
+
+        private void OnAudioFormatChanged(int sampleRate, int channelCount)
+        {
+            lock (this)
+            {
+                _audioClip = null;
+                // TODO Race condition here. AudioSource.clip still refers to the old clip. We can't modify it (not main thread).
+                // Thus we may get some calls to _audioData.OnClipRead (which is expecting old samplerate/channelcount)
+                // before the new clip is created.
+                EnqueueCreateAudioClipOnMainThread(sampleRate, channelCount);
+            }
+        }
+
+        private void RemoteAudioFrameReady(AudioFrame frame)
+        {
+            // The two parts of getting WebRTC audio:
+            //
+            // _audioData buffers the audioframes which arrive via WebRTC callbacks.
+            // WebRTC will sometimes change the format midstream, and if so, we update to match the new
+            // format immediately and will need to create a new AudioClip instance with the new parameters.
+            // Note we can't change AudioClip immediately since this is not the unity main thread.
+            //
+            // _audioClip is the unity object which can be attached to various audio emitters.
+            // The audioclip in this class is created with callbacks which read the data from _audioBuffer.
+            if (_audioClipEnqueuedCreate)
+            {
+                return; // waiting for creation.
```

Comment:
Can't we start buffering inside a pure C# object while waiting to create the Unity wrapper (`AudioClip`)? It seems we will take some latency on start here that we could avoid.

Reply:
New API hides this detail